domain/domain.go

```diff
@@ -932,6 +940,91 @@ func (do *Domain) initLogBackup(ctx context.Context, pdClient pd.Client) error {
 	return nil
 }
 
+// when tidb_replica_read = 'closest-adaptive', check tidb and tikv's zone label matches.
+// if not match, disable replica_read to avoid uneven read traffic distribution.
+func (do *Domain) closestReplicaReadCheckLoop(ctx context.Context, pdClient pd.Client) {
+	defer util.Recover(metrics.LabelDomain, "closestReplicaReadCheckLoop", nil, false)
+
+	// trigger check once instantly.
+	if err := do.checkReplicaRead(ctx, pdClient); err != nil {
+		logutil.BgLogger().Warn("refresh replicaRead flag failed", zap.Error(err))
+	}
+
+	ticker := time.NewTicker(time.Minute)
+	defer func() {
+		ticker.Stop()
+		do.wg.Done()
+		logutil.BgLogger().Info("closestReplicaReadCheckLoop exited.")
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := do.checkReplicaRead(ctx, pdClient); err != nil {
+				logutil.BgLogger().Warn("refresh replicaRead flag failed", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (do *Domain) checkReplicaRead(ctx context.Context, pdClient pd.Client) error {
+	// fast path
+	do.sysVarCache.RLock()
+	replicaRead := do.sysVarCache.global[variable.TiDBReplicaRead]
+	do.sysVarCache.RUnlock()
+
+	if !strings.EqualFold(replicaRead, "closest-adaptive") {
+		logutil.BgLogger().Info("closest replica read is not enabled, skip check!", zap.String("mode", replicaRead))
+		return nil
+	}
+	stores, err := pdClient.GetAllStores(ctx, pd.WithExcludeTombstone())
+	if err != nil {
+		return err
+	}
+
+	zones := make(map[string]int)
+	for _, s := range stores {
+		// skip tumbstone stores or tiflash
+		if s.NodeState == metapb.NodeState_Removing || s.NodeState == metapb.NodeState_Removed || engine.IsTiFlash(s) {
+			continue
+		}
+		for _, label := range s.Labels {
+			if label.Key == placement.DCLabelKey && label.Value != "" {
+				zones[label.Value] = 0
+				break
+			}
+		}
+	}
+
+	servers, err := infosync.GetAllServerInfo(ctx)
+	if err != nil {
+		return err
+	}
+	matches := true
+	for _, s := range servers {
+		if v, ok := s.Labels[placement.DCLabelKey]; ok && v != "" {
+			if _, ok := zones[v]; !ok {
+				matches = false
+				break
+			}
+			zones[v] += 1
+		}
+	}
+	if matches {
+		for _, count := range zones {
+			if count == 0 {
+				matches = false
+				break
+			}
+		}
+	}
+	if variable.SetEnableAdaptiveReplicaRead(matches) {
```

Comment:
What's the difference between `closest-adaptive` and `closest-read` after the PR https://github.com/pingcap/tidb/pull/35927?  Do them has the same dispatch strategy?

Reply:
The only different is that `closest-read` always try to dispatch all request to the closest store while `closest-adaptive` only dispatch request to the closest store when the estimate response size exceeds certain threshold and fallback other requests to region leader.